File: elixir/lib/2021/02_dive/aoc202102.ex
defmodule AOC2021.Day02 do
  @moduledoc """
  Advent of Code 2021, day 2: Dive!
  """
  require AOC

  @doc """
  Parse input
  """
  def parse(puzzle_input) do
    puzzle_input |> String.split("\n") |> Enum.map(&parse_command/1)
  end

  @doc """
  Parse a single command into a tuple

  ## Example:

      iex> parse_command("up 12")
      {:up, 12}
  """
  def parse_command("forward " <> steps), do: {:forward, steps |> String.to_integer()}
  def parse_command("up " <> steps), do: {:up, steps |> String.to_integer()}
  def parse_command("down " <> steps), do: {:down, steps |> String.to_integer()}

  @doc """
  Solve part 1
  """
  def part1(input) do
    input |> Enum.reduce({0, 0}, &move/2) |> Tuple.to_list() |> Enum.product()
  end

  @doc """
  Carry out one (part 1) move command

  ## Examples:

      iex> move({:forward, 12}, {10, 20})
      {22, 20}

      iex> move({:up, 3}, {10, 20})
      {10, 17}

      iex> move({:down, 14}, {10, 20})
      {10, 34}
  """
  def move({:forward, steps}, {horizontal_pos, depth}), do: {horizontal_pos + steps, depth}
  def move({:up, steps}, {horizontal_pos, depth}), do: {horizontal_pos, depth - steps}
  def move({:down, steps}, {horizontal_pos, depth}), do: {horizontal_pos, depth + steps}

  @doc """
  Solve part 2
  """
  def part2(input) do
    input
    |> Enum.reduce({0, 0, 0}, &move_with_aim/2)
    |> Tuple.to_list()
    |> Enum.slice(0..1)
    |> Enum.product()
  end

  @doc """
  Carry out one (part 2) move with aim command

  ## Examples:

      iex> move_with_aim({:up, 3}, {10, 20, 3})
      {10, 20, 0}

      iex> move_with_aim({:down, 14}, {10, 20, 3})
      {10, 20, 17}

      iex> move_with_aim({:forward, 12}, {10, 20, 3})
      {22, 56, 3}
  """
  def move_with_aim({:forward, steps}, {horizontal_pos, depth, aim}),
    do: {horizontal_pos + steps, depth + steps * aim, aim}

  def move_with_aim({:up, steps}, {horizontal_pos, depth, aim}),
    do: {horizontal_pos, depth, aim - steps}

  def move_with_aim({:down, steps}, {horizontal_pos, depth, aim}),
    do: {horizontal_pos, depth, aim + steps}

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/src/2021/02_dive/aoc202102.py
"""AoC 2, 2021: Dive!"""

# Standard library imports
import pathlib
import sys


def parse_data(puzzle_input):
    """Parse input."""
    return [parse_command(command) for command in puzzle_input.split("\n")]


def parse_command(command):
    """Parse one command.

    >>> parse_command("forward 12")
    (12, 0)

    >>> parse_command("up 3")
    (0, -3)

    >>> parse_command("down 7")
    (0, 7)
    """
    match command.split():
        case ["forward", steps]:
            return (int(steps), 0)
        case ["up", steps]:
            return (0, -int(steps))
        case ["down", steps]:
            return (0, int(steps))


def part1(data):
    """Solve part 1."""
    return sum(dhpos for dhpos, _ in data) * sum(ddepth for _, ddepth in data)


def part2(data):
    """Solve part 2."""
    horizontal_pos, depth, aim = 0, 0, 0
    for command in data:
        match command:
            case (0, daim):
                aim += daim
            case (dpos, 0):
                horizontal_pos += dpos
                depth += dpos * aim

    return horizontal_pos * depth


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().strip())
        print("\n".join(str(solution) for solution in solutions))
